### PR TITLE
Migrate to Vite!

### PR DESCRIPTION
**Pull request by gregory-j-baker**
### Migrate to Vite!

[Vite](https://vitejs.dev/) will eventually become the default compiler for Remix. To ensure that we have an upgrade path going forward, it is in our best interest to migrate.

## Description of changes

- removal of `express.server.ts` (functionality now moved to `entry.server.tsx` and various middlewares)
- updated `dockerfile` (now uses the vite build output)
- removal of the no-unused packages in `package.json`
- everything from the [Remix Vite migration guide](https://remix.run/docs/en/main/guides/vite#migrating)

### Checklist

- [x] I have tested the changes locally
- [x] I have added/updated tests that prove my fix is effective or that my feature works
- [x] I have checked that my code follows the project's coding style by running `npm run format:check`
- [x] I have checked that my code contains no linting errors by running `npm run lint`
- [x] I have checked that my code contains no type errors by running `npm run typecheck`
- [x] I have checked that all unit tests pass by running `npm run test:unit -- run`
- [x] I have checked that all e2e tests pass by running `npm run test:e2e`

### CODE REVIEW

File: frontend/vite.config.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    host: true,
+  },
+  plugins: [
+    envOnly(),
+    expressDevServer(),
+    remix({
+      ignoredRouteFiles: ['**/*'], // we will manually configure routes
+      routes: (defineRoutes) => jsonRoutes(defineRoutes, JSON.parse(readFileSync('./app/routes.json', 'utf8'))),
+      future: {
+        unstable_singleFetch: true,
```
Comment (sebastien-comeau) [inline frontend/vite.config.ts:66]: 🤷 
> This is an unstable API and will continue to change, do not adopt in production
https://remix.run/docs/en/main/guides/single-fetch#single-fetch

File: frontend/vitest.config.ts [hunk 2]
```diff
@@ -2,15 +2,16 @@
...
 
 import react from '@vitejs/plugin-react';
+import envOnly from 'vite-env-only';
 import tsconfigPaths from 'vite-tsconfig-paths';
 import { configDefaults, defineConfig } from 'vitest/config';
 
 // install global node polyfills
 // see: https://remix.run/docs/en/main/other-api/node#polyfills
-installGlobals();
+installGlobals({ nativeFetch: true });
```
Comment (sebastien-comeau) [inline frontend/vitest.config.ts:11]: Does it mean we can remove `undici` as dependency?

File: frontend/vite.config.ts (hunk 1 above)
Comment (gregory-j-baker) [reply frontend/vite.config.ts:66]: Hmm.. it is required for middleware..

<https://github.com/kiliman/remix-express-vite-plugin/blob/main/packages/remix-create-express-app/README.md#middleware>

File: frontend/vitest.config.ts (hunk 2 above)
Comment (gregory-j-baker) [reply frontend/vitest.config.ts:11]: Not sure. I was going to look into that after migration..

<https://remix.run/docs/en/main/guides/single-fetch#enabling-single-fetch>

**Comment by gregory-j-baker** [comment]
Gonna mark this as draft while I explore options for the unstable fetch.

File: frontend/app/entry.server.tsx [hunk 3]
```diff
@@ -17,11 +17,17 @@ import { getLocale, initI18n } from '~/utils/locale-utils.server';
 import { getLogger } from '~/utils/logging.server';
 import { randomHexString } from '~/utils/string-utils';
 
+export { expressApp as app } from '~/express.server';
```
Comment (sebastien-comeau) [inline frontend/app/entry.server.tsx:20]: Should add documentation for this line. I'm not sure why it's needed.

**Review by sebastien-comeau — APPROVED** [review]

**Review by sebastien-comeau — APPROVED** [review]

**Pull request merged**